### PR TITLE
Introduce new CycleScheduler class

### PR DESCRIPTION
**Pull request by KipK**
- add CycleScheduler for centralized multi-underlying cycle management
  Implement CycleScheduler class that orchestrates all cycles for multiple underlyings within a single master cycle. 
- new multiple underlyings power distribution algorithm 
- Integrate CycleScheduler into ThermostatSwitch, ThermostatValve and BaseThermostat
- Route TPI handler through CycleScheduler
- Valve: passthrough mode calling set_valve_open_percent() directly
- Move calculate_cycle_times() from timing_utils.py into cycle_scheduler.py.
  CycleScheduler now owns min_activation_delay and min_deactivation_delay as permanent attributes and computes 
  on_time_sec/off_time_sec internally in start_cycle().
- removed deprecated cycle_manager.py and timing_utils.py 

- Includes new unit tests for compute_offsets, lifecycle, callbacks, and edge cases (0%, 100%, HVAC OFF, force restart).
- Reworked previous pytest for new cycleScheduler logic.
- Added technical doc documentation/cycle_scheduler.md

Technical documentation: 
en: https://github.com/KipK/versatile_thermostat/blob/cycle_scheduler/documentation/tech-docs/cycle_scheduler.md
fr: https://github.com/KipK/versatile_thermostat/blob/cycle_scheduler/documentation/tech-docs/cycle_scheduler-fr.md

### CODE REVIEW

**Comment by KipK** [comment]
@jmcollin78 Ca fonctionne ici en test depuis hier. 
Comme tu vois ci dessous, cycleManager essaye de décaler les cycles ( en rouge )  sur la totalité de la fenetre, plutôt qu'à la suite. 
Ca permet de mieux répartir la puissance sur la durée du cycle
image

Le seul cas où plusieurs underlyings démarrent en meme temps sont les départs à 100% de puissance comme la fenêtre est pleine. 

On peut prevoir un param de minimal_delai dans le config_flow pour ce cas là si vraiment encore besoin.
Avec une logique du type, si le cycle précedent était sur off, alors faire démarrer les radiateurs en décalé avec ce délai. Mais ca ne sera plus un cycle à 100% du coup. Seul le premier rad aura vraiment fait 100%.


Les tests sont ok. 
Pour test_multiple_switch.py‎ j'ai du supprimer une partie qui devient obsolète.
Le test dédié au cycle_manager rempli cette fonction. Y a peut etre d'autre tests à faire dedans.

**Comment by jmcollin78** [comment]
Très bien. Comment ça marche avec 4 ? Tu peux le simuler ?

4 radiateurs à 80% :
Capture d’écran 2026-02-18 à 21 35 20

**Comment by KipK** [comment]
> Très bien. Comment ça marche avec 4 ? Tu peux le simuler ?
> 
Dans  la doc je t'ai mis un example avec 3 rads. Le principe est le même.

Je viens de decouvrir seulement maintenant qu'avec le dev container je peux lancer un HA de test en mode debug. :P 
Je vais me faire une config avec 4 switchs pour simuler ça. 

Quand je pense que je test sur ma prod depuis des mois, n'importe quoi...

**Comment by jmcollin78** [comment]
> Je viens de decouvrir seulement maintenant qu'avec le dev container je peux lancer un HA de test en mode debug. :P
Je vais me faire une config avec 4 switchs pour simuler ça.

Ah ben oui y a déjà tout de dispo avec des configs et tout. C'est comme ça que je teste en fait.

Faut tu personnalises un peu  le devcontainer, y a des chemins d'accès aux cartes et aux .ssh pour github si tu as. Sinon tu commentes.

**Comment by KipK** [comment]
Je bosse déjà sur le dev container depuis un moment en plus. 
J'avais pas remarqué qu'il y avait déjà un ha complet. 

~~Comment tu fais pour simuler la dernière vue du capteur pour pas que ca vienne m'embeter ?~~

**Comment by KipK** [comment]
image
avec 4 rads.
Comme tu vois ca calcule l'offset pour utiliser 100% de la fenetre.


Y a plus un test qui utilise les timings qui passent sur github ce soir.

**Comment by jmcollin78** [comment]
ok, je vois. J'ai l'impression que cette méthode de répartition de la charge lisse beaucoup moins bien que celle implémentée, c'est pour ça que tu pensais y arriver sans décalage.

Des fois tu as 4 rad actifs et des fois 1 seul:


Avec la méthode courante tu as toujours 3 ou 4 rad actifs:
Capture d’écran 2026-02-19 à 06 58 35

Je vais refaire avec un peu moins de puissance (60%) pour qu'on puisse comparer

**Comment by jmcollin78** [comment]
<je repasse sur discord>

**Comment by KipK** [comment]
Longue discussion aujourd'hui sur le sujet sur le discord. 
@bontiv ontiv à peut etre une meilleure solution qui répondrait aux 2 problématiques en meme temps, je le laisse proposer l'algo kivabien on va tester.

En attendant l'actuel et surtout le scheduler fait qd meme déjà le job proprement. C'est déjà ça :)

image

**Comment by jmcollin78** [comment]
Je vais regarder. J'ai fixé les tests au fait.

**Comment by KipK** [comment]
L'avant dernier commit a un peu foiré le merge conflict avec le PR de autotpi continuous  Kext. 

Faut que je le refasse, ou j'attende que tu merge ce PR auto-tpi avant.
Peut etre mieux que j'attende, car il y a des modifis necessaires ici pour ca aussi vu que c'est dépendant des cycles.

Le truc vraiment cool, maintenant on a des vrais callbacks de debut et fin de cycles propres. Et les algos register eux memes ces callbacks. La partie superieure devient complètement algo agnostic ( en preparation de pluginisation future )

**Comment by KipK** [comment]
C'est tout bon. Tu peux check. 

image


Les tests ok aussi. Y en avait un paquet à adapter tu vas voir :)

**Comment by jmcollin78** [comment]
T'as les github actions checkout qui failed.

**Comment by KipK** [comment]
> T'as les github actions checkout qui failed.

C'était un soucis côté GitHub, j'ai relancé le validate ça passe.

File: custom_components/versatile_thermostat/thermostat_prop.py [hunk 1]
```diff
@@ -277,15 +277,15 @@ async def async_set_auto_tpi_mode(
...
+        need to know the details.
         """
-        await self._fire_cycle_start_callbacks(
-            params.get("on_time_sec", 0),
-            params.get("off_time_sec", 0),
-            params.get("on_percent", 0),
-            params.get("hvac_mode", "stop")
-        )
+        self._cycle_scheduler = scheduler
+        if self._algo_handler and hasattr(self._algo_handler, "on_scheduler_ready"):
```
Comment (jmcollin78) [inline custom_components/versatile_thermostat/thermostat_prop.py]: C'est pour quoi ce 'hasattr(self._algo_handler, "on_scheduler_ready")' ? J'arrive pas à comprendre.

File: commit_message.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Fix CycleScheduler racollage computation and edge cases
```
Comment (jmcollin78) [inline commit_message.txt]: To be deleted, I guess.

File: fix_mocks.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import os
```
Comment (jmcollin78) [inline fix_mocks.py]: to be deleted ?

File: test_debug.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
+import asyncio
```
Comment (jmcollin78) [inline test_debug.py]: To be deleted ?

File: custom_components/versatile_thermostat/thermostat_prop.py (hunk 1 above)
Comment (KipK) [reply custom_components/versatile_thermostat/thermostat_prop.py]: C'est pour les algos qui ont besoin de s'enregistrer auprès du scheduler. 
_algo_handler peut etre soit PropHandlerTPI , soit par ex PropHandlerSmartPI.

TPI n'a pas besoin, il est purement proportionnel sans notion de cycles, il n'a pas de methode on_scheduler_ready. ( AutoTPI lui s'enregistre de son coté pour avoir le callback ) 

SmartPi  lui doit s'enregistrer. On appelle le callback uniquement si le handler l'expose.

File: commit_message.txt (hunk 2 above)
Comment (KipK) [reply commit_message.txt]: yep, remaining wastes. I'll remove them

File: custom_components/versatile_thermostat/thermostat_prop.py (hunk 1 above)
Comment (jmcollin78) [reply custom_components/versatile_thermostat/thermostat_prop.py]: C'est un peu crado ça comme façon de faire non ? Vaudrait pas mieux d'avoir une méthode `on_scheduler_ready` qui ne fait rien quand il n'y a rien à faire (comme toutes les autres dans ce cas là) ?

**Comment by jmcollin78** [comment]
Les tests sont ko à cause du portage sur homeassistant==2026.3.1. J'attends qu'il la release et ça devrait fixer les tests. Je vois qu'il y a eu pas mal de fix après ma relecture, je peux merge ou tu as d'autres choses en vue ?

**Comment by KipK** [comment]
non j'ai juste rebase sur ton main aujourd'hui, on est bon sur le scheduler 

edit: je viens de squash les commits.

**Comment by jmcollin78** [comment]
Please rebase, then yuou will have Python 3.14 for github actions which is needed for HA 2026.3.1.
Tests should pass now.

**Comment by jmcollin78** [comment]
You have 2 deprecation warning, you can fix in cycleScheduler:

/home/runner/work/versatile_thermostat/versatile_thermostat/custom_components/versatile_thermostat/cycle_manager.py:96: DeprecationWarning: 'asyncio.iscoroutinefunction' is deprecated and slated for removal in Python 3.16; use inspect.iscoroutinefunction() instead

**Comment by KipK** [comment]
> you have 2 deprecation warning, you can fix in cycleScheduler:

> /home/runner/work/versatile_thermostat/versatile_thermostat/custom_components/versatile_thermostat/cycle_manager.py:9


cycle_manager.py est supprimé dans ce PR, remplacé par le scheduler.

File: custom_components/versatile_thermostat/thermostat_prop.py (hunk 1 above)
Comment (KipK) [reply custom_components/versatile_thermostat/thermostat_prop.py]: oui on peut le supprimer. De toute façon je dis des bêtises, prop_handler_tpi.py a bien la methode on_scheduler_ready ( vu que c'est là qu'auto tpi register le callback ) 
C'est plus une sécu dans le cas d'autres algos qui oublieraient la méthode, mais je peux retirer.

Comment (jmcollin78) [reply custom_components/versatile_thermostat/thermostat_prop.py]: Yes vire stp. A chaque fois que je vais tomber dessus, je vais me reposer la question.

Comment (KipK) [reply custom_components/versatile_thermostat/thermostat_prop.py]: ok c'est fait.

**Comment by KipK** [comment]
Je me suis permis d'ajouter une petite sécurité pour rendre ca plus robuste dans le dernier commit.
edit: je le remet en draft e temps de tester. J'ai identifié un comportement bizarre.
